server/src/main/java/org/candlepin/model/ProductCurator.java

```diff
@@ -334,15 +273,346 @@ else if (boolAttrs != null && boolAttrs.contains(attr.getName())) {
+        // As an added bonus, it's quicker, but we'll have to be mindful of the memory vs backend
+        // state divergence.
+
+        // TODO:
+        // These may end up needing to change if we hit the MySQL 65k-element limitation on IN.
+
+        Session session = this.currentSession();
+        Set<String> ownerIds = new HashSet<String>();
+
+        for (Owner owner : owners) {
+            ownerIds.add(owner.getId());
+        }
+
+        // Owner products
+        String sql = "UPDATE cp2_owner_products SET product_uuid = ?1 " +
+            "WHERE product_uuid = ?2 AND owner_id IN (?3)";
+
+        int opCount = session.createSQLQuery(sql)
+            .setParameter("1", updated.getUuid())
+            .setParameter("2", current.getUuid())
+            .setParameterList("3", ownerIds)
+            .executeUpdate();
+
+        log.debug("{} owner-product relations updated", opCount);
+
+        // Activation key products
+        sql = "UPDATE cp2_activation_key_products SET product_uuid = ?1 " +
+            "WHERE product_uuid = ?2 AND key_id IN (SELECT id FROM cp_activation_key WHERE owner_id IN (?3))";
+
+        int akCount = session.createSQLQuery(sql)
+            .setParameter("1", updated.getUuid())
+            .setParameter("2", current.getUuid())
+            .setParameterList("3", ownerIds)
+            .executeUpdate();
+
+        log.debug("{} activation keys updated", akCount);
+
+        // Installed products
+        sql = "UPDATE cp2_installed_products SET product_uuid = ?1 " +
+            "WHERE product_uuid = ?2 AND consumer_id IN (SELECT id FROM cp_consumer WHERE owner_id IN (?3))";
+
+        int ipCount = session.createSQLQuery(sql)
+            .setParameter("1", updated.getUuid())
+            .setParameter("2", current.getUuid())
+            .setParameterList("3", ownerIds)
+            .executeUpdate();
+
+        log.debug("{} installed products updated", ipCount);
+
+        // pool provided and derived provided products
+        sql = "UPDATE cp_pool SET product_uuid = ?1 WHERE product_uuid = ?2 AND owner_id IN (?3)";
+
+        int ppCount = session.createSQLQuery(sql)
+            .setParameter("1", updated.getUuid())
+            .setParameter("2", current.getUuid())
+            .setParameterList("3", ownerIds)
+            .executeUpdate();
+
+        log.debug("{} pools updated", ppCount);
+
+        sql = "UPDATE cp_pool SET derived_product_uuid = ?1 " +
+            "WHERE derived_product_uuid = ?2 AND owner_id IN (?3)";
+
+        int pdpCount = session.createSQLQuery(sql)
+            .setParameter("1", updated.getUuid())
+            .setParameter("2", current.getUuid())
+            .setParameterList("3", ownerIds)
+            .executeUpdate();
+
+        log.debug("{} pools updated", pdpCount);
+
+        sql = "UPDATE cp2_pool_provided_products SET product_uuid = ?1 " +
+            "WHERE product_uuid = ?2 AND pool_id IN (SELECT id FROM cp_pool WHERE owner_id IN (?3))";
+
```

Comment:
Same IN concerns for selecting pool ids -- bonus pools can drive the pool counts up rather high as we've seen in some of the customer cases we've seen as of late.